diff --git a/plug/coldplug.c b/plug/coldplug.c
index 5fcb9a3..ed5018b 100644
--- a/plug/coldplug.c
+++ b/plug/coldplug.c
@@ -41,6 +41,9 @@ static void udevtrigger_complete(struct uloop_process *proc, int ret)
 void procd_coldplug(void)
 {
 	char *argv[] = { "udevtrigger", NULL };
+	/* Why are /dev unmounted and then remounted?!?!
+	 * It destroys what has already been populated. */
+#if 0
 	unsigned int oldumask = umask(0);
 
 	umount2("/dev/pts", MNT_DETACH);
@@ -50,6 +53,7 @@ void procd_coldplug(void)
 	mkdir("/dev/pts", 0755);
 	umask(oldumask);
 	mount("devpts", "/dev/pts", "devpts", MS_NOEXEC | MS_NOSUID, 0);
+#endif
 	udevtrigger.cb = udevtrigger_complete;
 	udevtrigger.pid = fork();
 	if (!udevtrigger.pid) {
